### PR TITLE
conv1d - added 1d convolution

### LINKED ISSUE #219: Support for conv1d
Currently `dfdx` supports `conv2d`, it would be if `conv1d` is supported out of the box. I know it'd be possible to use conv2d as a workaround but the code would not look clean.

### PR DESCRIPTION
**Pull request by jcrist1**
This PR only adds support for the conv1d operator based on [Issue #219](https://github.com/coreylowman/dfdx/issues/219). The implementation was heavy on copy paste, and I have glossed over the index calculations. Also I was unable to verify the cuda versions, as I don't have an Nvidia card. I have an idea where we could use property based testing and PyO3 to compare with pytorch in a more rigorous way.

For GH: Resolves #219

### CODE REVIEW

File: src/tensor_ops/conv1d/mod.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        lhs: &Tensor<L, E, Self>,
+        grad_lhs: &mut Self::Vec,
+        rhs: &Tensor<R, E, Self>,
+        grad_rhs: &mut Self::Vec,
+        out: &impl Tensorlike<O, E, Self>,
+        grad_out: &Self::Vec,
+    ) -> Result<(), Self::Err>;
+}
+
+/// Apply the 1d convolution to a tensor.
```
Comment (nkoppel) [inline src/tensor_ops/conv1d/mod.rs:52]:
```suggestion
/// Applies a 1d convolution to a tensor.
```

File: src/nn/conv1d.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    Const<{ I / G }>: Sized,
+    Conv1D<I, O, K, S, P, L, G, E, D>: BuildModule<D, E>,
+{
+    type Built = Conv1D<I, O, K, S, P, L, G, E, D>;
+    fn try_build_on_device(device: &D) -> Result<Self::Built, <D>::Err> {
+        Self::Built::try_build(device)
+    }
+}
+
+/// **Requires Nightly** Performs *unbiased* 2d convolutions on 3d and 4d images.
```
Comment (nkoppel) [inline src/nn/conv1d.rs:44]: Documentation needs an update here.

File: src/tensor_ops/conv1d/mod.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        self,
+        stride: Stride,
+        padding: Padding,
+        dilation: Dilation,
+        groups: Groups,
+    ) -> Self::Convolved {
+        self.try_conv1d(stride, padding, dilation, groups).unwrap()
+    }
+
+    /// Fallibly applies a 2D convolution to the input tensor.
```
Comment (nkoppel) [inline src/tensor_ops/conv1d/mod.rs:105]: Documentation needs updates.

**Review by nkoppel — COMMENTED** [review]
This looks really good except for a few places where you forgot to update documentation. @coreylowman will need to verify this on CUDA before merging.

**Comment by jcrist1** [comment]
@nkoppel addressed the documentation

File: src/tensor_ops/conv1d/cudnn_kernel.rs
Comment (coreylowman) [inline src/tensor_ops/conv1d/cudnn_kernel.rs:1]: cuDNN doesn't support 1d convolutions so we can remove this file

**Comment by jcrist1** [comment]
After a long weekend away from the computer I'm back. I've managed to try out the compile on lambda labs, and it failed. Will dig into it a bit to see if I can figure it out.

**Comment by jcrist1** [comment]
Results of
```sh
cargo test -F cuda
```

```
test result: ok. 366 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 18.52s

test result: ok. 187 passed; 0 failed; 4 ignored; 0 measured; 0 filtered out; finished in 4.07s
```

**Comment by jcrist1** [comment]
Okay it's quite broken:
```
failures:
    nn::conv1d::tests::test_grouped_forward_sizes
    tensor_ops::conv1d::tests::test_batched_conv1d
    tensor_ops::conv1d::tests::test_conv1d_default_stride_and_padding
    tensor_ops::conv1d::tests::test_conv1d_dilated
    tensor_ops::conv1d::tests::test_conv1d_grouped_forward
    tensor_ops::conv1d::tests::test_conv1d_padding_1
    tensor_ops::conv1d::tests::test_conv1d_s4p3k2
    tensor_ops::conv1d::tests::test_conv1d_stride_2
    tensor_ops::conv1d::tests::test_conv1d_stride_3_padding_4

test result: FAILED. 5 passed; 9 failed; 0 ignored; 0 measured; 409 filtered out; finished in 1.35s

error: test failed, to rerun pass `--lib`
ubuntu@146-235-237-134:~/dfdx$ failures:
446166, 0.72817594, 1.2460675, 0.2969481, 1.1754416, -0.57176435, 0.29306656, -0.19717735, -0.072016776, -0.24861556, 1.3089604, 0.35397184, -0.55793697]
out data [1.7531477, 0.44691145, 0.44691145, 0.10388159, 0.1279889, 0.1279889, -0.9886375, -0.66274095, -0.66274095], (Const, Const)
thread 'tensor_ops::conv1d::tests::test_conv1d_s4p3k2' panicked at 'lhs != rhs | 1.7531477 != 0.44691145', src/tensor_ops/conv1d/tests.rs:175:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tensor_ops::conv1d::tests::test_conv1d_padding_1 stdout ----
thread 'tensor_ops::conv1d::tests::test_conv1d_padding_1' panicked at 'lhs != rhs | 0.13320218 != 0.06176047', src/tensor_ops/conv1d/tests.rs:90:5

---- tensor_ops::conv1d::tests::test_conv1d_grouped_forward stdout ----
thread 'tensor_ops::conv1d::tests::test_conv1d_grouped_forward' panicked at 'lhs != rhs | -0.44480872 != -0.35965547', src/tensor_ops/conv1d/tests.rs:262:9

---- tensor_ops::conv1d::tests::test_conv1d_default_stride_and_padding stdout ----
thread 'tensor_ops::conv1d::tests::test_conv1d_default_stride_and_padding' panicked at 'lhs != rhs | -0.12854128 != 0.2249364', src/tensor_ops/conv1d/tests.rs:23:5

---- nn::conv1d::tests::test_grouped_forward_sizes stdout ----
thread 'nn::conv1d::tests::test_grouped_forward_sizes' panicked at 'called `Result::unwrap()` on an `Err` value: CublasError(CUBLAS_STATUS_INVALID_VALUE)', src/tensor_ops/conv1d/cuda_kernel.rs:126:22

---- tensor_ops::conv1d::tests::test_conv1d_stride_3_padding_4 stdout ----
thread 'tensor_ops::conv1d::tests::test_conv1d_stride_3_padding_4' panicked at 'lhs != rhs | 0.04796273 != -0.31665158', src/tensor_ops/conv1d/tests.rs:136:5

---- tensor_ops::conv1d::tests::test_conv1d_stride_2 stdout ----
thread 'tensor_ops::conv1d::tests::test_conv1d_stride_2' panicked at 'lhs != rhs | -0.16069691 != -0.03002486', src/tensor_ops/conv1d/tests.rs:55:5

---- tensor_ops::conv1d::tests::test_conv1d_dilated stdout ----
thread 'tensor_ops::conv1d::tests::test_conv1d_dilated' panicked at 'lhs != rhs | 0 != 1', src/tensor_ops/conv1d/tests.rs:225:5

---- tensor_ops::conv1d::tests::test_batched_conv1d stdout ----
```

**Comment by coreylowman** [comment]
Thanks for testing - and yeah debugging cuda code is a pain 😅

**Comment by jcrist1** [comment]
Sorry this fell by the wayside, had a lot going on the last two weeks. Will try to dig into this now

**Comment by coreylowman** [comment]
No worries at all, real life happens a lot in the open source community 😁

**Comment by coreylowman** [comment]
@jcrist1 I'm going to merge this into a conv1d branch so we can collab on it, I'll start pushing to that branch. Trying to get this implemented so I can test out a whisper implementation

**Comment by jcrist1** [comment]
@coreylowman thanks, I'm mostly throwing spaghetti at the wall trying to figure out why the buffers don't work. There's a lot of garbage now especially in the cuda code with println debugging

**Pull request merged**